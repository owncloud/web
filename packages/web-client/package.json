{
  "name": "@ownclouders/web-client",
  "version": "11.3.1",
  "description": "ownCloud web client",
  "license": "AGPL-3.0",
  "private": false,
  "type": "module",
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-client",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-client"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "default": "./dist/web-client.js",
        "require": "./dist/web-client.cjs",
        "types": "./dist/src/index.d.ts"
      },
      "./graph": {
        "default": "./dist/web-client/graph.js",
        "require": "./dist/web-client/graph.cjs",
        "types": "./dist/src/graph/index.d.ts"
      },
      "./graph/generated": {
        "default": "./dist/web-client/graph/generated.js",
        "require": "./dist/web-client/graph/generated.cjs",
        "types": "./dist/src/graph/generated/index.d.ts"
      },
      "./ocs": {
        "default": "./dist/web-client/ocs.js",
        "require": "./dist/web-client/ocs.cjs",
        "types": "./dist/src/ocs/index.d.ts"
      },
      "./sse": {
        "default": "./dist/web-client/sse.js",
        "require": "./dist/web-client/sse.cjs",
        "types": "./dist/src/sse/index.d.ts"
      },
      "./webdav": {
        "default": "./dist/web-client/webdav.js",
        "require": "./dist/web-client/webdav.cjs",
        "types": "./dist/src/webdav/index.d.ts"
      }
    }
  },
  "exports": {
    ".": {
      "default": "./src/index.ts"
    },
    "./graph": {
      "default": "./src/graph/index.ts"
    },
    "./graph/generated": {
      "default": "./src/graph/generated/index.ts"
    },
    "./ocs": {
      "default": "./src/ocs/index.ts"
    },
    "./sse": {
      "default": "./src/sse/index.ts"
    },
    "./webdav": {
      "default": "./src/webdav/index.ts"
    },
    "./l10n": {
      "default": "./l10n/translations.json"
    }
  },
  "scripts": {
    "generate-openapi": "rm -rf src/graph/generated && docker run --rm -v \"${PWD}/src/graph:/local\" openapitools/openapi-generator-cli generate -i https://raw.githubusercontent.com/owncloud/libre-graph-api/main/api/openapi-spec/v1.0.yaml  -g typescript-axios -o /local/generated",
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && rm -rf package/dist/tests && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "dependencies": {
    "@casl/ability": "^6.7.1",
    "@microsoft/fetch-event-source": "^2.0.1",
    "@ownclouders/web-pkg": "workspace:*",
    "axios": "^1.7.7",
    "fast-xml-parser": "^4.5.0",
    "lodash-es": "^4.17.21",
    "luxon": "^3.5.0",
    "uuid": "^11.0.0",
    "webdav": "^5.7.1",
    "xml-js": "^1.6.11",
    "zod": "^3.23.8"
  },
  "devDependencies": {
    "@types/luxon": "3.4.2",
    "@vitejs/plugin-vue": "5.1.4",
    "clean-publish": "5.0.0",
    "vite-plugin-dts": "4.2.4",
    "vite-plugin-node-polyfills": "0.22.0",
    "vite": "5.4.8"
  }
}
