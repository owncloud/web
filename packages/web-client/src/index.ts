import { graph } from './graph'
import { ocs } from './ocs'
import { webdav } from './webdav'

export * from './errors'
export * from './helpers'
export * from './utils'
export * from './constants'

export { graph, ocs, webdav }
