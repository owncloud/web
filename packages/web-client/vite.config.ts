import { join, resolve } from 'path'
import { defineConfig, searchForWorkspaceRoot } from 'vite'
import dts from 'vite-plugin-dts'
import { nodePolyfills } from 'vite-plugin-node-polyfills'

const projectRootDir = searchForWorkspaceRoot(process.cwd())

export default defineConfig({
  resolve: {
    alias: {
      crypto: join(projectRootDir, 'polyfills/crypto.js')
    }
  },
  build: {
    lib: {
      entry: {
        'web-client': resolve(__dirname, 'src/index.ts'),
        'web-client/graph': resolve(__dirname, 'src/graph/index.ts'),
        'web-client/graph/generated': resolve(__dirname, 'src/graph/generated/index.ts'),
        'web-client/ocs': resolve(__dirname, 'src/ocs/index.ts'),
        'web-client/sse': resolve(__dirname, 'src/sse/index.ts'),
        'web-client/webdav': resolve(__dirname, 'src/webdav/index.ts')
      }
    }
  },
  plugins: [
    nodePolyfills({
      exclude: ['crypto']
    }),
    dts({
      include: ['src'],
      outDir: 'dist/types',
      insertTypesEntry: true
    })
  ]
})
