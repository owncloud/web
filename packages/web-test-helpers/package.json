{
  "name": "@ownclouders/web-test-helpers",
  "version": "11.3.2",
  "description": "ownCloud web test helpers",
  "license": "AGPL-3.0",
  "private": false,
  "type": "module",
  "main": "src/index.ts",
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-test-helpers",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-test-helpers"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "import": "./dist/web-test-helpers.es.js",
        "require": "./dist/web-test-helpers.umd.cjs",
        "types": "./dist/src/index.d.ts"
      }
    }
  },
  "scripts": {
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "peerDependencies": {
    "@ownclouders/web-pkg": "workspace:^",
    "@vue/test-utils": "^2.4.6",
    "vue": "^3.5.10"
  },
  "devDependencies": {
    "@vitejs/plugin-vue": "5.1.4",
    "clean-publish": "^5.0.0",
    "vite-plugin-dts": "^4.2.3",
    "vite": "^5.4.8"
  },
  "dependencies": {
    "@casl/ability": "^6.7.1",
    "@casl/vue": "^2.2.2",
    "@ownclouders/design-system": "workspace:^",
    "@ownclouders/web-client": "workspace:^",
    "@pinia/testing": "^0.1.3",
    "axios": "1.7.7",
    "pinia": "2.2.6",
    "vitest-mock-extended": "2.0.2",
    "vue-router": "4.2.5",
    "vue3-gettext": "2.4.0"
  }
}
