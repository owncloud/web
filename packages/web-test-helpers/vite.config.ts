import { resolve } from 'path'
import { defineConfig } from 'vite'
import dts from 'vite-plugin-dts'
import pkg from './package.json' assert { type: 'json' }
import vue from '@vitejs/plugin-vue'

export default defineConfig({
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-test-helpers',
      fileName: (format) => `web-test-helpers.${format}.js`
    },
    rollupOptions: {
      external: [...Object.keys(pkg.dependencies), ...Object.keys(pkg.peerDependencies)]
    }
  },
  plugins: [vue(), dts({ include: ['src'], outDir: 'dist/types', insertTypesEntry: true })]
})
