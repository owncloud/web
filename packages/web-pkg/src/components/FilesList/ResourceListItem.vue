<template>
  <div
    class="oc-resource oc-text-overflow"
    :class="{ 'oc-resource-no-interaction': !isResourceClickable }"
  >
    <resource-link
      v-if="isIconDisplayed"
      :resource="resource"
      :link="link"
      :is-resource-clickable="isResourceClickable"
      class="oc-resource-link"
      @click="emitClick"
    >
      <oc-img
        v-if="hasThumbnail"
        :key="thumbnail"
        v-oc-tooltip="tooltipLabelIcon"
        :src="thumbnail"
        class="oc-resource-thumbnail"
        width="40"
        height="40"
        :aria-label="tooltipLabelIcon"
      />
      <resource-icon
        v-else
        v-oc-tooltip="tooltipLabelIcon"
        :aria-label="tooltipLabelIcon"
        :resource="resource"
      />
    </resource-link>
    <div class="oc-resource-details oc-text-overflow" :class="{ 'oc-pl-s': isIconDisplayed }">
      <resource-link
        :resource="resource"
        :is-resource-clickable="isResourceClickable"
        :link="link"
        class="oc-text-overflow"
        @click="emitClick"
      >
        <resource-name
          :key="resource.name"
          :name="resource.name"
          :path-prefix="pathPrefix"
          :extension="resource.extension"
          :type="resource.type"
          :full-path="resource.path"
          :is-path-displayed="isPathDisplayed"
          :is-extension-displayed="isExtensionDisplayed"
        />
      </resource-link>
      <div class="oc-resource-indicators">
        <component
          :is="parentFolderComponentType"
          v-if="isPathDisplayed"
          v-oc-tooltip="parentFolderNameTooltip"
          :to="parentFolderLink"
          :style="parentFolderStyle"
          class="parent-folder oc-text-truncate"
        >
          <oc-icon v-bind="parentFolderLinkIconAttrs" />
          <span class="text oc-text-truncate" v-text="parentFolderName" />
        </component>
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { Resource } from '@ownclouders/web-client'
import ResourceIcon from './ResourceIcon.vue'
import ResourceLink from './ResourceLink.vue'
import ResourceName from './ResourceName.vue'
import { RouteLocationRaw } from 'vue-router'

/**
 * Displays a resource together with the resource type icon or thumbnail
 */
export default defineComponent({
  name: 'ResourceListItem',
  components: { ResourceIcon, ResourceLink, ResourceName },
  props: {
    /**
     * The resource to be displayed
     */
    resource: {
      type: Object as PropType<Resource>,
      required: true
    },
    /**
     * The prefix that will be shown in the path
     */
    pathPrefix: {
      type: String,
      required: false,
      default: ''
    },
    /**
     * The resource link
     */
    link: {
      type: Object as PropType<RouteLocationRaw>,
      required: false,
      default: null
    },
    /**
     * Asserts whether the resource path should be displayed
     */
    isPathDisplayed: {
      type: Boolean,
      required: false,
      default: false
    },
    /**
     * The resource parent folder link path
     */
    parentFolderLink: {
      type: Object,
      required: false,
      default: null
    },
    /**
     * The resource parent folder name to be displayed
     */
    parentFolderName: {
      type: String,
      required: false,
      default: ''
    },
    /**
     * The resource parent folder link path icon additional attributes
     */
    parentFolderLinkIconAdditionalAttributes: {
      type: Object,
      required: false,
      default: () => ({})
    },
    /**
     * Asserts whether the resource extension should be displayed
     */
    isExtensionDisplayed: {
      type: Boolean,
      required: false,
      default: true
    },
    /**
     * Asserts whether the resource thumbnail should be displayed
     */
    isThumbnailDisplayed: {
      type: Boolean,
      required: false,
      default: true
    },
    /**
     * Asserts whether the resource thumbnail should be displayed
     */
    isIconDisplayed: {
      type: Boolean,
      required: false,
      default: true
    },
    /**
     * Asserts whether clicking on the resource name triggers any action
     */
    isResourceClickable: {
      type: Boolean,
      required: false,
      default: true
    }
  },
  emits: ['click'],
  computed: {
    parentFolderNameTooltip() {
      return this.parentFolderName
    },
    parentFolderComponentType() {
      return this.parentFolderLink ? 'router-link' : 'span'
    },

    parentFolderStyle() {
      return {
        cursor: this.parentFolderLink ? 'pointer' : 'default'
      }
    },

    parentFolderLinkIconAttrs() {
      return {
        'fill-type': 'line',
        name: 'folder-2',
        size: 'small',
        ...this.parentFolderLinkIconAdditionalAttributes
      }
    },

    hasThumbnail() {
      return (
        this.isThumbnailDisplayed &&
        Object.prototype.hasOwnProperty.call(this.resource, 'thumbnail')
      )
    },

    thumbnail() {
      return this.resource.thumbnail
    },

    tooltipLabelIcon() {
      if (this.resource.locked) {
        return this.$gettext('This item is locked')
      }
      return null
    }
  },

  methods: {
    emitClick() {
      /**
       * Triggered when the resource is a file and the name is clicked
       */
      this.$emit('click')
    }
  }
})
</script>

<style lang="scss">
.oc-resource {
  align-items: center;
  display: inline-flex;
  justify-content: flex-start;
  overflow: visible !important;

  &-no-interaction {
    pointer-events: none;
  }

  &-link {
    position: relative;
  }

  &-thumbnail {
    border-radius: 2px;
    object-fit: cover;
    height: $oc-size-icon-default * 1.5;
    max-height: $oc-size-icon-default * 1.5;
    width: $oc-size-icon-default * 1.5;
    max-width: $oc-size-icon-default * 1.5;
  }

  &-details {
    display: block;

    a {
      text-decoration: none;
    }

    a:hover,
    a:focus {
      outline-offset: 0;
    }
  }

  &-indicators {
    display: flex;

    a {
      &:hover {
        background-color: var(--oc-color-input-bg);
        border-radius: 2px;
      }

      .text {
        &:hover {
          color: var(--oc-color-text-default);
          text-decoration: underline;
        }
      }
    }

    .parent-folder {
      display: flex;
      align-items: center;

      padding: 0 2px 0 2px;
      margin: 0 8px 0 -2px;

      .oc-icon {
        padding-right: 3px;
      }

      .text {
        font-size: 0.8125rem;
        color: var(--oc-color-text-muted);
      }
    }
  }
}
</style>
