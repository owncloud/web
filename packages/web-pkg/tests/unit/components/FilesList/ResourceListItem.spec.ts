import { defaultComponentMocks, defaultPlugins, mount } from '@ownclouders/web-test-helpers'
import ResourceListItem from '../../../../src/components/FilesList/ResourceListItem.vue'
import { HIDDEN_FILE_EXTENSIONS, Resource } from '@ownclouders/web-client'
import { mock } from 'vitest-mock-extended'

const SELECTORS = Object.freeze({
  resourceName: '.oc-resource-basename'
})

const fileResource = {
  name: 'forest.jpg',
  path: 'nature/forest.jpg',
  thumbnail: 'https://cdn.pixabay.com/photo/2015/09/09/16/05/forest-931706_960_720.jpg',
  type: 'file',
  isFolder: false,
  extension: 'jpg'
} as Resource
const folderResource = {
  name: 'Documents',
  path: '',
  type: 'folder',
  isFolder: true
} as Resource
const fileResourceWithoutParentFoldername = {
  name: 'example.pdf',
  path: 'example.pdf',
  type: 'file',
  isFolder: false,
  extension: 'pdf'
} as Resource

describe('OcResource', () => {
  it("doesn't emit a click if the resource is a folder", () => {
    const wrapper = mount(ResourceListItem, {
      props: {
        resource: folderResource,
        targetRoute: {
          name: 'tests-route'
        }
      },
      global: {
        stubs: { RouterLink: true },
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()]
      }
    })

    wrapper.find('.oc-resource-name').trigger('click')
    expect(wrapper.emitted('click')).toBeFalsy()
  })

  it("doesn't emit a click if the resource is not clickable", () => {
    const wrapper = mount(ResourceListItem, {
      props: {
        resource: fileResource,
        isResourceClickable: false
      },
      global: {
        stubs: { RouterLink: true },
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()]
      }
    })

    wrapper.find('.oc-resource-name').trigger('click')
    expect(wrapper.emitted('click')).toBeFalsy()
  })

  it('emits a click', async () => {
    const wrapper = mount(ResourceListItem, {
      props: {
        resource: fileResource
      },
      global: {
        stubs: { RouterLink: true },
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()]
      }
    })

    await wrapper.find('.oc-resource-name').trigger('click')
    expect(wrapper.emitted('click')).toHaveLength(1)
  })

  it('parent folder component type is link if parent folder given', () => {
    const wrapper = mount(ResourceListItem, {
      props: {
        resource: fileResource,
        isPathDisplayed: true,
        parentFolderLink: {}
      },
      global: {
        stubs: { RouterLink: true },
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()]
      }
    })

    expect(wrapper.find('.parent-folder').exists()).toBeTruthy()
    expect(wrapper.find('.parent-folder').attributes('style')).toEqual('cursor: pointer;')
  })

  it('parent folder component type is span if parent folder not given', () => {
    const wrapper = mount(ResourceListItem, {
      props: {
        resource: fileResource,
        isPathDisplayed: true
      },
      global: {
        stubs: { RouterLink: true },
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()]
      }
    })

    expect(wrapper.find('.parent-folder').find('a').exists()).toBeFalsy()
    expect(wrapper.find('.parent-folder').attributes('style')).toEqual('cursor: default;')
  })

  it('displays parent folder name default if calculated name is empty', () => {
    const wrapper = mount(ResourceListItem, {
      props: {
        resource: fileResourceWithoutParentFoldername,
        isPathDisplayed: true,
        parentFolderName: 'Example parent folder name'
      },
      global: {
        stubs: { RouterLink: true },
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be used without icon/thumbnail', () => {
    const wrapper = mount(ResourceListItem, {
      props: {
        resource: fileResourceWithoutParentFoldername,
        isIconDisplayed: false,
        parentFolderName: 'Example parent folder name'
      },
      global: {
        stubs: { RouterLink: true },
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins()]
      }
    })

    expect(wrapper.find('oc-resource-thumbnail').exists()).toBeFalsy()
    expect(wrapper.find('oc-resource-icon').exists()).toBeFalsy()
  })

  it.each(HIDDEN_FILE_EXTENSIONS)(
    'should not show the file extension when it equals "%s"',
    (extension) => {
      const { wrapper } = getWrapper({
        props: {
          resource: mock<Resource>({ extension, name: 'forest.' + extension }),
          isExtensionDisplayed: true
        }
      })

      expect(wrapper.find(SELECTORS.resourceName).text()).toStrictEqual('forest')
    }
  )
})

function getWrapper({ props }: { props: { resource: Resource; isExtensionDisplayed?: boolean } }) {
  const mocks = defaultComponentMocks()

  return {
    mocks,
    wrapper: mount(ResourceListItem, {
      props,
      global: {
        mocks,
        provide: mocks,
        plugins: [...defaultPlugins()]
      }
    })
  }
}
