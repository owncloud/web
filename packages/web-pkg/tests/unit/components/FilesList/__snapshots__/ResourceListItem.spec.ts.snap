// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OcResource > displays parent folder name default if calculated name is empty 1`] = `
"<div class="oc-resource oc-text-overflow"><button target="_self" type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-undefined oc-button-passive oc-button-passive-raw oc-resource-link" draggable="false">
    <!--v-if-->
    <!-- @slot Content of the button --> <span class="oc-icon oc-icon-l oc-icon-passive oc-resource-icon oc-resource-icon-file"><!----></span>
  </button>
  <div class="oc-resource-details oc-text-overflow oc-pl-s"><button target="_self" type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-left oc-button-gap-undefined oc-button-passive oc-button-passive-raw oc-text-overflow" draggable="false">
      <!--v-if-->
      <!-- @slot Content of the button --> <span class="oc-resource-name" data-test-resource-path="example.pdf" data-test-resource-name="example.pdf" data-test-resource-type="file"><span class="oc-text-truncate"><span class="oc-resource-basename">example</span></span><span class="oc-resource-extension">.pdf</span></span>
    </button>
    <div class="oc-resource-indicators"><span style="cursor: default;" class="parent-folder oc-text-truncate"><span class="oc-icon oc-icon-s oc-icon-passive"><!----></span> <span class="text oc-text-truncate">Example parent folder name</span></span></div>
  </div>
</div>"
`;
